File: web/src/lib/api/messages.ts
import { logger } from '../logger';
import { getApiUrl, TAG, AuthError, getToken, getDeviceAccessToken, isAuthFailure, withAuthRetry } from './client';
import { DeviceSendRateLimitedError, publishDeviceSendQuota, readQuotaFromResponse } from './deviceSendQuota';

export type MessageEnvelope = {
  type: 'text' | 'file' | 'control' | 'lan_file_offer' | 'lan_pull_probe' | 'lan_pull_probe_result'
    | 'lan_http_probe' | 'lan_http_probe_result' | 'lan_pull_cancelled'
    | 'webrtc_probe' | 'webrtc_probe_result'
    | 'webrtc_offer' | 'webrtc_answer' | 'webrtc_ice_candidate' | 'webrtc_transfer_cancel'
    | 'device_pair_hello'
    | 'device_roster_patch';
  payload: unknown;
  fromDeviceId: string;
  ts: number;
  id?: number;
  toDeviceId?: string;
  threadKey?: string;
};

/** Local-only fields for optimistic UI and progress */
export type LocalStatus = 'sending' | 'uploading' | 'downloading' | 'sent' | 'failed' | 'cancelled';
export type ChatMessage = MessageEnvelope & {
  _localId?: string;
  _status?: LocalStatus;
  _progress?: number;
  _speed?: string;
};

export type MessageHistoryItem = MessageEnvelope & { id: number };

export async function getMessageHistory(limit = 50, before?: number, threadKey?: string): Promise<MessageHistoryItem[]> {
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const params = new URLSearchParams({ limit: String(limit) });
    if (before != null && before > 0) params.set('before', String(before));
    if (threadKey != null && threadKey !== '') params.set('threadKey', threadKey);
    const res = await fetch(`${getApiUrl()}/api/messages/history?${params}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'getMessageHistory failed', res.status);
      throw new Error('Failed to load history');
    }
    const list = (await res.json()) as MessageHistoryItem[];
    logger.info(TAG, 'getMessageHistory success count=', list.length);
    return list;
  });
}

export async function deleteMessage(id: number): Promise<void> {
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const res = await fetch(`${getApiUrl()}/api/messages/${id}`, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'deleteMessage failed', res.status);
      throw new Error('Failed to delete message');
    }
    logger.debug(TAG, 'deleteMessage ok');
  });
}

export async function deleteThreadMessages(threadKey: string): Promise<void> {
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('Not authenticated');
    const params = new URLSearchParams({ threadKey });
    const res = await fetch(`${getApiUrl()}/api/messages/thread?${params}`, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'deleteThreadMessages failed', res.status);
      throw new Error('Failed to clear messages');
    }
    logger.debug(TAG, 'deleteThreadMessages ok');
  });
}

export async function sendMessage(data: MessageEnvelope): Promise<void> {
  logger.info(TAG, 'sendMessage type=', data.type, 'fromDeviceId=', data.fromDeviceId);
  const userToken = getToken();
  if (userToken) {
    return withAuthRetry(async () => {
      const token = getToken();
      if (!token) throw new Error('Not authenticated');
      const res = await fetch(`${getApiUrl()}/api/messages/send`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
        body: JSON.stringify({ data }),
      });
      if (isAuthFailure(res)) throw new AuthError();
      if (!res.ok) {
        logger.warn(TAG, 'sendMessage failed', res.status);
        throw new Error('Failed to send message');
      }
      logger.debug(TAG, 'sendMessage ok');
    });
  }
  const deviceToken = getDeviceAccessToken();
  if (!deviceToken) throw new Error('Not authenticated');
  const res = await fetch(`${getApiUrl()}/api/messages/device-send`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${deviceToken}` },
    body: JSON.stringify({ data }),
  });
  const quota = await readQuotaFromResponse(res);
  if (quota) publishDeviceSendQuota(quota);
  if (res.status === 429) {
    logger.warn(TAG, 'sendMessage device-send 429');
    throw new DeviceSendRateLimitedError(quota ?? { message: { used: 90, limit: 90, remaining: 0, retryAfterMs: 1000 }, signaling: { used: 0, limit: 600, remaining: 600, retryAfterMs: 0 }, kind: 'message', limited: true });
  }
  if (!res.ok) {
    logger.warn(TAG, 'sendMessage device-send failed', res.status);
    throw new Error('Failed to send message');
  }
  logger.debug(TAG, 'sendMessage device-send ok');
}

export async function pairDevice(peerDeviceId: string): Promise<void> {
  const deviceToken = getDeviceAccessToken();
  if (!peerDeviceId || !deviceToken) return;
  const res = await fetch(`${getApiUrl()}/api/devices/pair`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${deviceToken}` },
    body: JSON.stringify({ peerDeviceId }),
  });
  if (!res.ok) {
    logger.warn(TAG, 'pairDevice failed', res.status);
  }
}

File: app/lib/api/messages.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../logger.dart';
import 'client.dart';
import 'device_send_quota.dart';

class MessageEnvelope {
  final String type;
  final dynamic payload;
  final String fromDeviceId;
  final int ts;
  final int? id;
  final String? localId;
  /// Optional: directed delivery / filtering for realtime.
  final String? toDeviceId;
  /// Canonical conversation key (see `chat/thread_key.dart`).
  final String? threadKey;

  MessageEnvelope({
    required this.type,
    required this.payload,
    required this.fromDeviceId,
    required this.ts,
    this.id,
    this.localId,
    this.toDeviceId,
    this.threadKey,
  });

  factory MessageEnvelope.fromJson(Map<String, dynamic> j) {
    final tsVal = j['ts'];
    final idVal = j['id'];
    final ts = tsVal is num
        ? tsVal.toInt()
        : int.tryParse(tsVal?.toString() ?? '') ?? 0;
    final id = idVal == null
        ? null
        : (idVal is num ? idVal.toInt() : int.tryParse(idVal.toString()));
    return MessageEnvelope(
      type: j['type']?.toString() ?? '',
      payload: j['payload'],
      fromDeviceId: j['fromDeviceId']?.toString() ?? '',
      ts: ts,
      id: id,
      toDeviceId: j['toDeviceId']?.toString(),
      threadKey: j['threadKey']?.toString(),
    );
  }
}

/// Fetches message history (newest first). [before] is message id for cursor.
Future<List<MessageEnvelope>> getMessageHistory({
  int limit = 50,
  int? before,
  String? threadKey,
}) async {
  logApi.info('getMessageHistory limit=$limit before=$before threadKey=$threadKey');
  return withAuthRetry(() async {
    final query = <String, String>{'limit': limit.toString()};
    if (before != null && before > 0) query['before'] = before.toString();
    if (threadKey != null && threadKey.isNotEmpty) {
      query['threadKey'] = threadKey;
    }
    final uri = Uri.parse(
      '$apiBaseUrl/api/messages/history',
    ).replace(queryParameters: query);
    final r = await http.get(uri, headers: apiHeaders);
    checkAuthResponse(r, fallback: '加载历史失败');
    final list = (jsonDecode(r.body) as List)
        .map((e) => MessageEnvelope.fromJson((e as Map<String, dynamic>)))
        .toList();
    logApi.info('getMessageHistory success count=${list.length}');
    return list;
  });
}

Future<void> deleteMessage(int messageId) async {
  logApi.info('deleteMessage id=$messageId');
  return withAuthRetry(() async {
    final r = await http.delete(
      Uri.parse('$apiBaseUrl/api/messages/$messageId'),
      headers: apiHeaders,
    );
    checkAuthResponse(r, fallback: '删除失败');
    logApi.fine('deleteMessage ok id=$messageId');
  });
}

Future<void> deleteThreadMessages(String threadKey) async {
  logApi.info('deleteThreadMessages threadKey=$threadKey');
  return withAuthRetry(() async {
    final uri = Uri.parse('$apiBaseUrl/api/messages/thread').replace(
      queryParameters: {'threadKey': threadKey},
    );
    final r = await http.delete(uri, headers: apiHeaders);
    checkAuthResponse(r, fallback: '清空消息失败');
    logApi.fine('deleteThreadMessages ok threadKey=$threadKey');
  });
}

Future<void> sendMessage(Map<String, dynamic> data) async {
  final type = data['type'];
  final fromDeviceId = data['fromDeviceId'];
  logApi.info('sendMessage type=$type fromDeviceId=$fromDeviceId');
  if (hasAccessToken) {
    return withAuthRetry(() async {
      final r = await http.post(
        Uri.parse('$apiBaseUrl/api/messages/send'),
        headers: apiHeaders,
        body: jsonEncode({'data': data}),
      );
      checkAuthResponse(r, fallback: '发送失败');
      if (r.statusCode != 204) {
        throw Exception(r.body.isNotEmpty ? r.body : '发送失败');
      }
      logApi.fine('sendMessage ok');
    });
  }
  final r = await http.post(
    Uri.parse('$apiBaseUrl/api/messages/device-send'),
    headers: deviceApiHeaders,
    body: jsonEncode({'data': data}),
  );
  publishDeviceSendQuotaFromResponse(r);
  if (r.statusCode == 429) {
    final quota = parseDeviceSendQuota(
      headers: r.headers,
      body: r.body,
      limited: true,
    );
    throw DeviceSendRateLimitedException(
      quota ??
          DeviceSendQuota.fromJson(const {
            'error': 'rate_limited',
            'kind': 'message',
          }, limited: true),
    );
  }
  if (r.statusCode != 204) {
    throw Exception(errorMessageFromResponse(r, '发送失败'));
  }
  logApi.fine('sendMessage device-send ok');
}

Future<void> pairDevice(String peerDeviceId) async {
  if (peerDeviceId.isEmpty || !hasDeviceAccessToken) return;
  final r = await http.post(
    Uri.parse('$apiBaseUrl/api/devices/pair'),
    headers: deviceApiHeaders,
    body: jsonEncode({'peerDeviceId': peerDeviceId}),
  );
  if (r.statusCode != 204 && r.statusCode != 200) {
    logApi.warning('pairDevice failed status=${r.statusCode}');
  }
}
